src/backend/distributed/transaction/lock_graph.c

```diff
@@ -662,9 +669,34 @@ IsProcessWaitingForSafeOperations(PGPROC *proc)
 	PROCLOCK *waitProcLock = proc->waitProcLock;
 	LOCK *waitLock = waitProcLock->tag.myLock;
 
-	return waitLock->tag.locktag_type == LOCKTAG_RELATION_EXTEND ||
-		   waitLock->tag.locktag_type == LOCKTAG_PAGE ||
-		   waitLock->tag.locktag_type == LOCKTAG_SPECULATIVE_TOKEN;
+
+	if (waitLock->tag.locktag_type == LOCKTAG_RELATION_EXTEND ||
+		waitLock->tag.locktag_type == LOCKTAG_PAGE ||
+		waitLock->tag.locktag_type == LOCKTAG_SPECULATIVE_TOKEN)
+	{
+		return true;
+	}
+# if PG_VERSION_NUM >= PG_VERSION_14
+
+	/*
+	 * We ignore processes that are waiting for a short time, to give the
```

Comment:
Hmm my assumption was that deadlock detection factor already provides this benefit, but probably that was a wrong assumption.

Though, the node that is doing the deadlock detection might have a different value than current `DistributedDeadlockDetectionTimeoutFactor/DeadlockTimeout`. Should we maybe pass a parameter to `dump_local_wait_edges /dump_global_wait_edges ` so that caller can decide on the offset such as `DistributedDeadlockDetectionTimeoutFactor * (double) DeadlockTimeout`?

Reply:
TBH, this part of the fix I'm least confident about and I think probably needs some more thinking. The issue you mentioned is one of the reasons, another reason is that I'm worried that due to postgres rearranging processes it might be possible that new transactions are placed in front of an old process that has been stuck for a long time. Ignoring those new edges in the wait graph could then maybe cause a distributed deadlock to not be detected for longer than we want.

Reply:
In any case I'll split this PR into multiple smaller ones for ease of review.